File: basic/scripts/gfxtext.py
# *******************************************************************************************
# *******************************************************************************************
#
#		Name : 		gfxtext.py
#		Purpose :	Convert graphics file to/from text
#		Date :		27th February 2024
#		Author : 	Paul Robson (paul@robsons.org.uk)
#
# *******************************************************************************************
# *******************************************************************************************

import sys,re

# *******************************************************************************************
#
#									Convertor Class
#
# *******************************************************************************************

class GraphicsConvert(object):
	#
	#		Convert .gfx format to text format
	#
	def toText(self,source,target):
		hOutput = open(target,"w")  																	# Open file, write header
		hOutput.write("#\n#\tConversion of \"{0}\"\n#\n".format(source))
		src = [x for x in open(source,"rb").read(-1)]  													# Read graphics file
		offset1 = 256 + src[1] * 16 * 16 // 2  															# sprite 16 start here
		offset2 = offset1 + src[2] * 16 * 16 // 2  														# sprite 32 start here
		self.xlat = ".abcdefghijklmno"
		hOutput.write("\n[Tile:16x16]\n")  																# export 16x16 tiles
		for i in range(0,src[1]):
			hOutput.write("\n[{0}]\n".format(i))
			self.export(hOutput,src,256 + i * 16 * 16 // 2,16,16)
		hOutput.write("\n[Sprite:16x16]\n")  															# export 16x16 sprites
		for i in range(0,src[2]):
			hOutput.write("\n[{0}]\n".format(i))
			self.export(hOutput,src,offset1 + i * 16 * 16 // 2,16,16)  
		hOutput.write("\n[Sprite:32x32]\n")  															# export 32x32 sprites
		for i in range(0,src[3]):
			hOutput.write("\n[{0}]\n".format(i))
			self.export(hOutput,src,offset2 + i * 32 * 32 // 2,32,32)

		hOutput.close()
		return self
	#
	def export(self,h,src,p,width,height):
		while (height > 0):  																			# for each line
			height = height - 1
			for i in range(0,width >> 1):  																# for each nibble pair
				b = src[p]  																			# read it
				p += 1
				h.write(self.xlat[(b >> 4)])  															# output halves
				h.write(self.xlat[(b & 0x0F)])
			h.write("\n")
	#
	#		Convert text format to .gfx format
	#
	def toGraphics(self,source,target):
		self.src = open(source).readlines()  															# read in file
		self.src = [x.strip().upper() for x in self.src if not (x.startswith("#") or x.strip() == "")]
		self.tiles16 = []  																				# objects collections
		self.sprites16 = []
		self.sprites32 = []
		self.collections = [self.tiles16, self.sprites16,self.sprites32 ]   							# collection of collections
		self.currentCollection = 0  																	# This one we are using.
		#
		while len(self.src) != 0:  																		# while more to do 
			cmd = self.src.pop(0)
			assert cmd.startswith("[") and cmd.endswith("]"),"No command found"  
			if cmd == "[TILE:16X16]":  																# graphic type selection
				self.currentCollection = 0
			elif cmd == "[SPRITE:16X16]":
				self.currentCollection = 1
			elif cmd == "[SPRITE:32X32]":
				self.currentCollection = 2
			elif cmd == "[]":  																			# [] is define next free
				self.addDefinition(len(self.collections[self.currentCollection]))  						
			else:
				m = re.match("^\\[(\\d+)\\]$",cmd)  													# Load in to specific slot
				assert m is not None,"Bad command "+cmd
				self.addDefinition(int(m.group(1)))  													# add that one in.				
		#
		self.render = [ 0 ] * 256  																		# create header
		self.render[0] = 1
		self.render[1] = len(self.tiles16)
		self.render[2] = len(self.sprites16)
		self.render[3] = len(self.sprites32)
		for t in self.tiles16:  																		# render all tiles and sprites
			self.renderSpriteTile(t,16)
		for t in self.sprites16:
			self.renderSpriteTile(t,16)
		for t in self.sprites32:
			self.renderSpriteTile(t,32)
		#
		h = open(target,"wb")  																			# Write out the result
		h.write(bytes(self.render))
		h.close()
		return self
	#
	def addDefinition(self,n):
		while len(self.collections[self.currentCollection]) <= n:  										# Pad out to required length with Nones
			self.collections[self.currentCollection].append(None)
		assert self.collections[self.currentCollection][n] is None,"Duplicate sprite/tile"    			# Check for non duplicate	
		size = 32 if self.currentCollection == 2 else 16  												# size of sprite/tile
		self.collections[self.currentCollection][n] = self.src[:size]  									# get data for sprite
		self.src = self.src[size:]
	#
	def renderSpriteTile(self,data,size):
		for d in data:  																				# for each line of data
			for c in range(0,size >> 1):  																# for each character pair
				self.render.append(self.convert(d[c*2]) * 16 + self.convert(d[c*2+1]))  				# convert to nibble format
	#
	def convert(self,c):
		c = "@" if c == "." else c.upper()  															# must now be A-O
		assert c >= '@' and c <= 'O',"Bad graphic data"
		return ord(c) - ord('@')

if __name__ == "__main__":
	GraphicsConvert().toText("graphics.gfx","graphics.txt").toGraphics("graphics.txt","_test.gfx")

File: basic/scripts/gfxtext.c
// *******************************************************************************************
// *******************************************************************************************
//
//		Name : 		gfxtext.c
//		Purpose :	Convert text file to graphics file
//		Date :		28th February 2024
//		Author : 	Paul Robson (paul@robsons.org.uk)
//
// *******************************************************************************************
// *******************************************************************************************

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

static uint8_t *graphMem;
static uint16_t maxSize,currentSize;
static char lineBuffer[128];

// *******************************************************************************************
//
//									Initialise the conversion
//
// *******************************************************************************************

static void _CTGInitialise(uint8_t *gMem,uint16_t size) {
	graphMem = gMem;maxSize = size;  													// Remember position and size
	currentSize = 256;  																// Allocated so far, just the header
	for (int i = 0;i < size;i++) graphMem[i] = 0;  										// Zero the lot.
	graphMem[0] = 1;  																	// Format ID marker.
}

// *******************************************************************************************
//
//									Get position of entry.
//
// *******************************************************************************************

static uint16_t _CTGGetPosition(uint8_t block,uint8_t id) {
	uint16_t addr = 0;
	switch(block) {
		case 1:
			addr = 256 + id * (16 * 16 / 2);
			break;
		case 2:
			addr = 256 + graphMem[1] * (16 * 16 / 2) + id * (16 * 16 / 2);
			break;
		case 3:
			addr = 256 + (graphMem[1]+graphMem[2]) * (16 * 16 / 2) + id * (32 * 32 / 2);
			break;
	}
	return addr;
}

// *******************************************************************************************
//
//									Insert new block into memory
//
// *******************************************************************************************

static int _CTGAppendBlock(uint8_t block) {
	uint16_t pos = _CTGGetPosition(block,graphMem[block]);  							// Where the new one will go.
	uint16_t required = (block == 3) ? 32*32/2 : 16*16/2;  								// Byte size.
	if (currentSize + required >= maxSize) return 1;
	// printf("@%x %d %d\n",pos,block,graphMem[block]);
	if (pos == currentSize) {  															// Appending.
		// printf("Appended.\n");
	} else {
		// printf("Insert %d at %x\n",required,pos);  										// Make space for it.
		uint16_t p = currentSize;  														// Presented out of order.
		while (p >= pos) {
			graphMem[p + required] = graphMem[p];
			p--;
		}
	}
	graphMem[block]++; 																	// One more block.
	currentSize += required;  															// Just make it bigger.
	return 0;
}

// *******************************************************************************************
//
//							Allocate a new block for a tile or sprite
//
// *******************************************************************************************

static int _CTGAllocate(uint8_t block,uint8_t id) {
	uint8_t maxObj = (block == 1) ? 128 : 64;  											// This many at most
	if (id >= maxObj) return 1;   														// Too many !
	while (graphMem[block] <= id) {  													// Insert blank blocks until required exists.
		int e = _CTGAppendBlock(block); 			 									// Until the current one is valid.
		if (e != 0) return e;  															// Or append failed (out of memory)
	}
	return 0;
}

// *******************************************************************************************
//
//									Read a line from the source
//
// *******************************************************************************************

static char *_CTGGetLine(FILE *f) {
	if (fgets(lineBuffer,sizeof(lineBuffer),f) == NULL) return NULL;  					// End of file.
	if (lineBuffer[0] == '#') return _CTGGetLine(f);  									// Comment
	while (lineBuffer[0] != '\0' && lineBuffer[strlen(lineBuffer)-1] <= ' ') { 			// Right strip
		lineBuffer[strlen(lineBuffer)-1] = '\0';
	}
	char *p = lineBuffer;  																// Capitalise.
	while (*p != '\0') {
		*p = toupper(*p);  																// Capitalise
		if (*p == '.') *p = '@';  														// . to A-1
		p++;
	}
	if (lineBuffer[0] == '\0') return _CTGGetLine(f);  									// Blank line.
	return lineBuffer;
}

// *******************************************************************************************
//
//											Load in a block
//
// *******************************************************************************************

static int _CTGLoadBlock(FILE *f,uint8_t block,uint8_t graphic) {
	int e = _CTGAllocate(block,graphic);  												// Check space available
	if (e != 0) return e; 
	uint8_t size = (block == 3) ? 32 : 16;  											// Object size.
	uint16_t pos = _CTGGetPosition(block,graphic);  									// Where it is in the memory block
	// printf("%d %d %d $%x\n",block,graphic,size,pos);
	for (int row = 0;row < size;row++) {  												// Read each line of graphic data
		char *line = _CTGGetLine(f);
		if (line == NULL) return 3;  													// EOF
		if (strlen(line) != size) return 3;  											// Line the wrong size
		for (int col = 0; col < size;col++) {  											// Each character
			if (line[col] < '@' || line[col] > 'O') return 4;   						// Invalid data
			if ((col & 1) == 0) {  														// First nibble
				graphMem[pos] = (line[col] & 0x0F) << 4;
			} else {  																	// Second nibble
				graphMem[pos] |= (line[col] & 0x0F);
				pos++;
			}
		}
	}
	return 0;
}

// *******************************************************************************************
//
//								Convert given file to graphics format
//
//							 returns size loaded, or < 256 if error occurs
//
// *******************************************************************************************

uint16_t ConvertTextToGraphics(FILE *source,uint8_t *graphicMemory,size_t size) {
	_CTGInitialise(graphicMemory,size);  												// Set up the buffer.
	uint8_t currentBlock = 1; 
	char *line;
	int isOk;
	while (line = _CTGGetLine(source), line != NULL) {  								// Scan through the source
		isOk = 0;
		if (strcmp(line,"[TILE:16X16]") == 0) isOk = currentBlock = 1;  				// Check for block switches.
		if (strcmp(line,"[SPRITE:16X16]") == 0) isOk = currentBlock = 2;
		if (strcmp(line,"[SPRITE:32X32]") == 0) isOk = currentBlock = 3;
		if (strcmp(line,"[]") == 0) {   												// Next in current block.
			int e = _CTGLoadBlock(source,currentBlock,graphMem[currentBlock]);
			if (e != 0) return e;  														// Creating that block failed.
			isOk = 1;
		}
		if (isOk == 0) {  																// Not processed
			if (line[0] != '[' || line[strlen(line)-1] != ']') return 2; 				// It must be [<n>]
			for (int i = 1;i < strlen(line)-1;i++) {
				if (!isdigit(line[i])) return 2;
			}
			line[strlen(line)-1] = '\0';  												// Strip ]
			int e = _CTGLoadBlock(source,currentBlock,atoi(line+1));   					// Load that numbered block graphic
			if (e != 0) return e;
		}	
	}
	return currentSize;
}

// *******************************************************************************************
//
//								Simple caller used for testing
//
// *******************************************************************************************

#ifdef NEO6502RUNTEST

static uint8_t buffer[32768];

int main() {
	FILE *f = fopen("graphics.txt","r");
	uint16_t size = ConvertTextToGraphics(f,buffer,sizeof(buffer));
	fclose(f);

	f = fopen("_cdemo.gfx","wb");
	fwrite(buffer,size,1,f);
	fclose(f);
	return 0;
}

#endif
